app/models/tag.rb

```diff
@@ -471,24 +471,35 @@ def autocomplete_prefixes
   end
 
   def add_to_autocomplete(score = nil)
-    score ||= autocomplete_score
-    if self.is_a?(Character) || self.is_a?(Relationship)
-      parents.each do |parent|
-        REDIS_AUTOCOMPLETE.zadd(self.transliterate("autocomplete_fandom_#{parent.name.downcase}_#{type.downcase}"), score, autocomplete_value) if parent.is_a?(Fandom)
-      end
+    parents.each do |parent|
+      add_to_fandom_autocomplete(parent, score)
     end
     super
   end
 
+  def add_to_fandom_autocomplete(fandom, score = nil)
+    return unless canonical
+    return unless fandom.is_a?(Fandom)
+    return unless self.is_a?(Character) || self.is_a?(Relationship)
+
+    score ||= autocomplete_score
+    REDIS_AUTOCOMPLETE.zadd(self.transliterate("autocomplete_fandom_#{fandom.name.downcase}_#{type.downcase}"), score, autocomplete_value)
+  end
+
   def remove_from_autocomplete
     super
-    if self.is_a?(Character) || self.is_a?(Relationship)
```

Comment:
If this is removed and shifted into       `remove_from_fandom_autocomplete` (and its add counterpart), this might lead to more iterations than necessary. Would an early return be a significant improvement?